Update clustering module docs

cbethell opened the pull request:
**Issue Addressed**
Closes #218

**What is the purpose of these changes?**
<!--Provide some background on the changes proposed.-->
The purpose of this PR is to add more detail to the clustering module `README.md` file on the multiple cluster type options, as well as the additional config file.

**What changes did you make?**
<!--Describe the concrete changes that you made or additional features that were added, be as detailed as possible in the steps you took.-->
This PR updates the `optional-clustering-analysis/README.md` with information on:
- the specific `cluster_config.yaml` file and the parameters in there
- the option to test and compare more than one clustering type at a time 

This PR also updates the config file paths where applicable in the clustering `README`, as well as in the main directory's `README`.

**Any comments, concerns, or questions important for reviewers**
- Are there any additional details you feel may be beneficial here before making this clustering module available publicly?

**Checklist**

Place an `x` in all boxes that you have completed.

- [ ] I have run the most recent version of the code
- [ ] If I am adding in reports or generating any plots, I have reviewed the necessary reports
- [x] I have added all necessary documentation (if applicable)

allyhawkins commented on README.md:
```diff
@@ -314,7 +319,18 @@ For example, if using the default values of Louvain clustering with a nearest ne
+
+For more on what's in the clustering analysis workflow and how to run the workflow, see the [`README.md`](optional-clustering-analysis/README.md#optional-clustering-analysis) file in the clustering analysis subdirectory.
+
+### The optional genes of interest analysis pipeline (In development)
```
Can you add a note below this heading that this is still in development and that updates will be coming soon. I just want to make sure that it's clear that this is not fully developed at this time.

allyhawkins commented on optional-clustering-analysis/README.md:
```diff
@@ -2,7 +2,15 @@
+
+We evaluate the clustering results across each of the parameters tested by looking at the following measures:
+
+1. **Cluster purity** - looks at each cell as the proportion of neighboring cells that are assigned to the same cluster, with a range of 0 to 1. Well-separated clusters should show little overlap between member and neighboring cells, and therefore high purity values for all member cells.
```
```suggestion
1. **Cluster purity** - The cluster purity is calculated for each cell as the proportion of neighboring cells that are assigned to the same cluster, with a range of 0 to 1. Well-separated clusters should show little overlap between member and neighboring cells, and therefore high purity values for all member cells.
```
Just adding some clarifying language here.

allyhawkins commented on optional-clustering-analysis/README.md:
```diff
@@ -2,7 +2,15 @@
+We evaluate the clustering results across each of the parameters tested by looking at the following measures:
+
+1. **Cluster purity** - looks at each cell as the proportion of neighboring cells that are assigned to the same cluster, with a range of 0 to 1. Well-separated clusters should show little overlap between member and neighboring cells, and therefore high purity values for all member cells.
+2. **Silhouette width** - looks at how well-separated each of the clusters are, with a range of -1 to 1. For each cell, the average distance to all cells in the same cluster and the average distance to all cells in another cluster, are calculated. The silhouette width for each cell is defined as the difference between these two values divided by their maximum. Therefore, cells will ideally have l ...
```
```suggestion
2. **Silhouette width** - Silhouette width is calculated for each cell across clusters and measures how well-separated each of the clusters are, with a range of -1 to 1. For each cell, the average distance to all cells in the same cluster and the average distance to all cells in another cluster, are calculated. The silhouette width for each cell is defined as the difference between these two values divided by their maximum distance. Therefore, cells will ideally have large positive silhouette widths, as this would mean that the cells of one cluster are well-separated from other clusters.
```

allyhawkins commented on optional-clustering-analysis/README.md:
```diff
@@ -2,7 +2,15 @@
+
+1. **Cluster purity** - looks at each cell as the proportion of neighboring cells that are assigned to the same cluster, with a range of 0 to 1. Well-separated clusters should show little overlap between member and neighboring cells, and therefore high purity values for all member cells.
+2. **Silhouette width** - looks at how well-separated each of the clusters are, with a range of -1 to 1. For each cell, the average distance to all cells in the same cluster and the average distance to all cells in another cluster, are calculated. The silhouette width for each cell is defined as the difference between these two values divided by their maximum. Therefore, cells will ideally have l ...
+3. **Cluster stability** - looks at the stability of the clustering results associated with each of the clustering parameters. Here, cells within each dataset are sampled using bootstrapping and the sampled cells are re-clustered. The new clustering assignments are compared to the original assignment by obtaining an adjusted rand index, and this process is repeated 20 times. Clustering results wi ...
```
```suggestion
3. **Cluster stability** - The stability of the clustering results associated with each of the clustering parameters tested is also reported. Here, cells within each dataset are sampled using bootstrapping and the sampled cells are re-clustered. The new clustering assignments are compared to the original assignment by obtaining an adjusted rand index, and this process is repeated 20 times. Clustering results with high stability would reveal that clustering after each bootstrap replicate is consistent with the original clustering results. We use summary adjusted Rand Index (ARI) values in the plots shown in the clustering report output by the workflow to represent the average calculated cluster stability values. The range here is 0 to 1, where stable clusters have values closer to 1.
```

allyhawkins commented on optional-clustering-analysis/README.md:
```diff
@@ -15,15 +23,15 @@ The clustering analysis workflow provided here can be used to explore different
+1. **Cluster Calculations**: clustering results are calculated for the provided type(s) of graph-based clustering, and range of nearest neighbor values.
 Cluster validity and stability results are also calculated.
 2. **Cluster Plots**: once clustering results are calculated and stored in the `SingleCellExperiment` object, the results from each of the clustering methods tested are displayed in a UMAP plot.
 Additionally, metrics associated with each of the clustering results such as silhouette width, cluster purity, and cluster stability are calculated and plotted.
```
```suggestion
Additionally, metrics associated with each of the clustering results such as silhouette width, cluster purity, and cluster stability (as described above) are calculated and plotted.
```

allyhawkins commented on optional-clustering-analysis/README.md:
```diff
@@ -40,28 +48,28 @@ To run this workflow, you will need to provide:
 5. A `_clustering_summary_stability_stats.tsv` file with the average adjusted Rand Index (ARI) values across each cluster assignment.
 
-The TSV files will be saved in a subdirectory called `clustering_stats/`.
+The TSV files will be saved in a subdirectory called `{library_id}_{filtering_method}_clustering_stats/`.
```
```suggestion
The TSV files will be saved in a subdirectory called `{library_id}_{filtering_method}_clustering_stats/`, where both the `{library_id}` and `{filtering_method}` are obtained from the input metadata file. 
```

allyhawkins commented on optional-clustering-analysis/README.md:
```diff
@@ -40,28 +48,28 @@ To run this workflow, you will need to provide:
 
 1. The RDS file containing the [output `SingleCellExperiment` object](../README.md#expected-output) from the core dowstream analyses workflow.
 2. The [`library_id`](../README.md#metadata-file-format) associated with the `SingleCellExperiment` object.
```
I think what we actually need here instead of `library_id` is the metadata file correct? Because we need the `library_id` and `filtering_method` that's read in from the metadata file? https://github.com/AlexsLemonade/scpca-downstream-analyses/blob/45955da9ecea4a4ac9ef4b591245f62acf0ce61a/cluster.snakefile#L7-L8

allyhawkins commented on optional-clustering-analysis/README.md:
```diff
@@ -89,15 +100,19 @@ snakemake --snakefile cluster.snakefile \
-### Processing parameters
+### Clustering parameters
+
+We have provided an additional [configuration file](https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html), `config/cluster_config.yaml` which sets the defaults for all parameters needed to run the clustering workflow.
```
```suggestion
We have provided an additional [configuration file](https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html), `config/cluster_config.yaml` which sets the defaults for all parameters needed to run the clustering workflow.
It is not required to alter these parameters to run the workflow, but if you would like to change the type of clustering or range of nearest neighbor parameters, this is done by changing these parameters. 
```
I'm not crazy about this exact wording, but I just want to emphasize that you don't need to use these parameters unless you want to explore other options so feel free to change this!

allyhawkins commented on optional-clustering-analysis/README.md:
```diff
@@ -89,15 +100,19 @@ snakemake --snakefile cluster.snakefile \
 
 | Parameter        | Description | Default value |
 |------------------|-------------|---------------|
+| `optional_cluster_types` | a comma-separated list of graph-based clustering type(s), values can be include "louvain" and/or "walktrap" (see more on these graph-based clustering methods in this [Community Detection Algorithms article](https://towardsdatascience.com/community-detection-algorithms-9bd8951e7dae)) | "louvain,walktrap" |
```
```suggestion
| `optional_cluster_types` | a comma-separated list of graph-based clustering type(s), options include "louvain" and/or "walktrap" (see more on these graph-based clustering methods in this [Community Detection Algorithms article](https://towardsdatascience.com/community-detection-algorithms-9bd8951e7dae)) | "louvain,walktrap" |
```

allyhawkins reviewed (commented):
I left a few minor comments throughout, but this looks pretty good. 

My one comment for the future that we don't need to think about quite yet... but as we start to add outputs that include already processed data to scpca-nf, we need to modify the snakefile to take as input a filepath similar to how we do for the core workflow so we don't require a specific file name. But that I think is an associated problem for later that we just need to be aware of. Other than that documentation and the workflow looks good for release prep once this PR is in.

cbethell commented:
Thanks fro the review @allyhawkins!! I've implemented your feedback and believe this is ready for another look now 👍

sjspielman commented on README.md:
```diff
@@ -197,11 +202,11 @@ Each library ID should have a unique `filepath`.
 ## Running the workflow
 
-We have provided an example [snakemake configuration file](https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html), [`config.yaml`](https://github.com/AlexsLemonade/scpca-downstream-analyses/blob/main/config.yaml) which sets the defaults for all parameters needed to run the workflow.
+We have provided an example [snakemake configuration file](https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html), [`config/config.yaml`](https://github.com/AlexsLemonade/scpca-downstream-analyses/blob/main/config/config.yaml) which sets the defaults for all parameters needed to run the workflow.
```
Do we want this to be a relative link to the config? Currently it's not targeted for the `main` branch so this link doesn't work and won't work after merge, either.

allyhawkins replied:
Just going to echo this statement that we want to update these links throughout as Stephanie suggested to be relative links to the files.

sjspielman commented on README.md:
```diff
@@ -211,7 +216,7 @@ These include the following parameters:
 | `mito_file` | full path to a file containing a list of mitochondrial genes specific to the genome or transcriptome version used for alignment. By default, the workflow will use the mitochondrial gene list obtained from Ensembl version 104 which can be found in the `reference-files` directory. |
 
-|[View Config File](https://github.com/AlexsLemonade/scpca-downstream-analyses/blob/main/config.yaml)|
+|[View Config File](https://github.com/AlexsLemonade/scpca-downstream-analyses/blob/main/config/config.yaml)|
```
Same comment re-relative link.

sjspielman commented on optional-clustering-analysis/README.md:
```diff
@@ -89,15 +100,20 @@ snakemake --snakefile cluster.snakefile \
 | `nearest_neighbors_increment` | the increment to use when implementing the range number of nearest neighbors for cluster stats (e.g. a value of 5 with min of 5 and max of 25 will test the nearest neighbors values of 5, 10, 15, 20, and 25) | 5 |
 | `overwrite_results` | a binary value indicating whether or not to overwrite any existing clustering results | `TRUE` |
 
+|[View Clustering Config File](https://github.com/AlexsLemonade/scpca-downstream-analyses/blob/development/config/cluster_config.yaml)|
```
This link is pointing to `development` branch, which will work once this is merged, but not when we merge that branch into `main`. So again, we probably want relative links.

sjspielman reviewed (commented):
This looks mostly fine! I've left some comments about where we might benefit from relative links to files.
Also, `doctoc` needs to be re-run for at least the `optional-clustering-analysis/README.md`. It would be good to generally make sure that README section links are working as well.

allyhawkins commented on optional-clustering-analysis/README.md:
```diff
@@ -15,18 +23,18 @@ The clustering analysis workflow provided here can be used to explore different
 There are two main steps of this clustering analysis workflow:
 
-1. **Cluster Calculations**: clustering results are calculated for the provided type of graph-based clustering, and range of nearest neighbor values.
+1. **Cluster Calculations**: clustering results are calculated for the provided type(s) of graph-based clustering, and range of nearest neighbor values.
```
```suggestion
1. **Cluster Calculations**: Clustering results are calculated for the provided type(s) of graph-based clustering, and range of nearest neighbor values.
```

allyhawkins commented on optional-clustering-analysis/README.md:
```diff
@@ -15,18 +23,18 @@ The clustering analysis workflow provided here can be used to explore different
-1. **Cluster Calculations**: clustering results are calculated for the provided type of graph-based clustering, and range of nearest neighbor values.
+1. **Cluster Calculations**: clustering results are calculated for the provided type(s) of graph-based clustering, and range of nearest neighbor values.
 Cluster validity and stability results are also calculated.
 2. **Cluster Plots**: once clustering results are calculated and stored in the `SingleCellExperiment` object, the results from each of the clustering methods tested are displayed in a UMAP plot.
```
```suggestion
2. **Cluster Plots**: Once clustering results are calculated and stored in the `SingleCellExperiment` object, the results from each of the clustering methods tested are displayed in a UMAP plot.
```

allyhawkins commented on optional-clustering-analysis/README.md:
```diff
@@ -39,29 +47,29 @@ If you are using conda, dependencies can be installed as [part of the initial se
 1. The RDS file containing the [output `SingleCellExperiment` object](../README.md#expected-output) from the core dowstream analyses workflow.
-2. The [`library_id`](../README.md#metadata-file-format) associated with the `SingleCellExperiment` object.
-2. The desired type of graph-based clustering (can be "louvain" or "walktrap"), along with the minimum, maximum, and incremental values that will be used to define the range of nearest neighbor values to be tested.
+2. A project metadata tab-separated value (TSV) file containing relevant information about your data necessary for processing, the same metadata file used in the core workflow can be used here (see more on this in the ["Metadata file format" section](../README.md#metadata-file-format) and an example of this metadata file [here](https://github.com/AlexsLemonade/scpca-downstream-analyses/blob/main/ ...
```
```suggestion
2. A project metadata tab-separated value (TSV) file containing relevant information about your data necessary for processing, the same metadata file used in the core workflow should be used here (see more on this in the ["Metadata file format" section](../README.md#metadata-file-format) and an example of this metadata file [here](https://github.com/AlexsLemonade/scpca-downstream-analyses/blob/main/project-metadata/example-library-metadata.tsv))
```
I think we want to be explicit that the exact metadata file path is required.

allyhawkins reviewed (commented):
A few more comments in addition to addressing the links that Stephanie brought up.

cbethell commented:
Okay I've incorporated the suggestions, updated the links to be relative links (please let me know if I missed any that you can see), re-ran doctoc and tested the links. Let me know what you both think now!

sjspielman commented on optional-clustering-analysis/README.md:
```diff
@@ -15,18 +23,18 @@ The clustering analysis workflow provided here can be used to explore different
+Additionally, metrics associated with each of the clustering results such as silhouette width, cluster purity, and cluster stability (as described above) are calculated and plotted.
 The plots are displayed in a html report for ease of reference.
 
 **Note** that the same [software requirements for the core workflow](https://github.com/AlexsLemonade/scpca-downstream-analyses/tree/development#3-additional-dependencies) are also required for this clustering workflow.
```
Missed this link first time around since it wasn't a diff in this PR, but we should make it relative too:

`https://github.com/AlexsLemonade/scpca-downstream-analyses/tree/development#3-additional-dependencies` --> `../README.md#3-additional-dependencies`

sjspielman commented on optional-clustering-analysis/README.md:
```diff
@@ -70,6 +78,9 @@ These include the following parameters:
 | `results_dir` | relative path to the directory where output files will be stored (use the same `results_dir` used in the prerequisite core workflow) |
 | `project_metadata` | relative path to your specific project metadata TSV file (use the same `project_metadata` used in the prerequisite core workflow) |
 
+|[View Config File](https://github.com/AlexsLemonade/scpca-downstream-analyses/blob/development/config/config.yaml)|
```
```suggestion
|[View Config File](../config/config.yaml)|
```

sjspielman commented on optional-clustering-analysis/README.md:
```diff
@@ -39,29 +47,29 @@ If you are using conda, dependencies can be installed as [part of the initial se
 1. The RDS file containing the [output `SingleCellExperiment` object](../README.md#expected-output) from the core dowstream analyses workflow.
-2. The [`library_id`](../README.md#metadata-file-format) associated with the `SingleCellExperiment` object.
-2. The desired type of graph-based clustering (can be "louvain" or "walktrap"), along with the minimum, maximum, and incremental values that will be used to define the range of nearest neighbor values to be tested.
+2. A project metadata tab-separated value (TSV) file containing relevant information about your data necessary for processing, the same metadata file used in the core workflow should be used here (see more on this in the ["Metadata file format" section](../README.md#metadata-file-format) and an example of this metadata file [here](https://github.com/AlexsLemonade/scpca-downstream-analyses/blob/ma ...
```
Noting we still have a full link here but should be fine since `main` branch.

cbethell replied:
Yeah I was torn on leaving it since it is pointing to the `main` branch but changed anyway in https://github.com/AlexsLemonade/scpca-downstream-analyses/pull/228/commits/48d67dabb360ebd4fb53b1e01a93987c5257067f!

sjspielman reviewed (approved):
Looks good! A couple small comments but I don't need to see again, unless you want to re-request me!

allyhawkins reviewed (approved):
LGTM 🚀
